### PR TITLE
Anonymous user segmentation

### PR DESCRIPTION
**Pull request by adambraimbridge**
Work in progress. Please don't merge.

### CODE REVIEW

File: server/middleware/set-allocation-id.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+var nodeUuid = require('node-uuid');
+
+// Set the allocation ID as an app variable.
+var getAllocationID = function(req, res){
+
+	// Requests receive an A/B allocation for each feature that has an active A/B test.
+	// A/B allocation is based on an allocation ID. Ideally the allocation ID is the
+	// user's uuid, but if the user is anonymous, it fails back to their device ID.
+	// See: http://git.svc.ft.com/projects/NEXT/repos/preflight-requests/browse/src/tasks/session.js
+	var sessionFailure = req.get('ft-preflight-session-failure') || req.get('x-ft-preflight-session-failure');
```
Comment (adambraimbridge) [inline server/middleware/set-allocation-id.js]: From Paul: In future an "FT-Session-Response" header will be available. So we'll know if the server's down. Possibly

File: server/middleware/set-allocation-id.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	var uuid = req.get('ft-user-id') || req.get('x-ft-user-id');
+	var sessionToken = req.get('ft-session-token') || req.get('x-ft-session-token');
+	var deviceID = req.get('ft-device-id') || req.get('x-ft-device-id');
+	var isAnonymous = req.get('ft-anonymous-user') || req.get('x-ft-anonymous-user');
```
Comment (adambraimbridge) [inline server/middleware/set-allocation-id.js]: From Paul: This is not really a useful header; it's just the opposite of FT-Session-Token

**Comment by commuterjoy** [comment]
Probably good to go, but we need to do the CDN work first and check we don't break beacon/spoor.

**Comment by commuterjoy** [comment]
this is now a mega commit. lets merge this now and pick up problems in fresh branches.

**Pull request merged**